src/pyk/proof/reachability.py

```
@@ -592,32 +578,48 @@ def __init__(
             priority=1,
         )
 
-        self._target_is_terminal = kcfg_explore.kcfg_semantics.is_terminal(
-            self.proof.kcfg.node(self.proof.target).cterm
-        )
+        self._target = self.proof.kcfg_exploration.kcfg.node(self.proof.target)
+        self._target_is_terminal = kcfg_explore._check_terminal(self._target)
+
+        # All nodes marked as terminal are semantically terminal
+        for node in self.proof.terminals:
+            assert kcfg_explore._check_terminal(node)
+
+        # Target node is not marked as terminal
+        assert self._target.id not in self.proof.kcfg_exploration._terminals
```

Also this? Why can we not have target node be terminal? It is actually usually the case.

---

Because the intended meaning of `_terminals` here is "terminal nodes obtained during execution". This excludes `target` - the check whether or not `target` is terminal is only interesting for whether or not the subsumption check is done on every obtained node or not.

---

`terminal` is a generic property that could apply to any node, not about how the node came about. It's just a mapping from nodes to bool.

We in fact _need_ `target` node to be `terminal` if we want to disable `implication_every_block` automatically.

---

The `is_terminal` function of a `KCFGSemantics` can be applied to any node, and it is applied to the `target` node when this check is needed. Specifically, in line 571 of `reachability.py`:
```
self._target_is_terminal = kcfg_explore._check_terminal(self._target)
```

I will check what would happen if `target` would be included.

---

I'm including `target` now and have removed the "correctness" checks

---

This leads to the `test_fail_fast` tests failing fast, because they are assuming the meaning that I originally intended:
```
 >       assert len(proof._terminal) == 2
E       assert 3 == 2
E        +  where 3 = len({2, 6, 7})
E        +    where {2, 6, 7} = <pyk.proof.reachability.APRProof object at 0x7fe088a4fac0>._terminal
```
where 7 is the target node. 

Shall I adjust these tests?

---

@ehildenb I think this is the only remaining bit of the PR to be corrected - I can either correct the passing tests to include the target node as terminal or exclude target from terminal nodes, as was previously done. Let me know which you'd prefer.

---

I think adjust the test, yes.

---

Very well